Address bugs in BigInteger

ts2do opened the pull request:
Method `Add(ref BigInteger lhs, uint value, ref BigInteger result)` would store most of the result blocks into `lhs` instead of `result`
Method `ShiftLeft(ulong input, uint shift, ref BigInteger output)` with a shift argument not evenly divisible by 32 would generally compute the higher blocks incorrectly

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -353,7 +353,7 @@ public static void Add(ref BigInteger lhs, uint value, ref BigInteger result)
                 {
                     ulong sum = (ulong)(lhs._blocks[index]) + carry;
-                    lhs._blocks[index] = (uint)(sum);
+                    result._blocks[index] = (uint)(sum);
```
👍, this isn't failing in production because we only ever call it as `Add(ref this, value, ref this);`

tannergooding commented:
Looking at the things being fixed in this PR and the item called out by Anton. These bugs don't cause issues in production because the code code paths/failure cases in question aren't currently hit (and I don't believe ever will be hit).

This type is only meant for (and is designed around) `float`/`double` formatting/parsing; so I think the appropriate way to fix these bugs would be to remove the unnecessary logic. That should both "fix" the bugs and would potentially make the code faster for its intended purpose.

For example, `public static void Add(ref BigInteger lhs, uint value, ref BigInteger result)` is only ever called as `public void Add(uint value) => Add(ref this, value, ref this);` and `public static void ShiftLeft(ulong input, uint shift, ref BigInteger output)` is only ever used as `ShiftLeft(1, exponent, ref result);` (where `result = new BigInteger(0)`).

tannergooding commented:
Also, for reference, it looks like some of these bugs existed in the original `BigNum.cpp` code that this code was ported from: <https://github.com/dotnet/coreclr/pull/19999>

AntonLapounov commented:
@tannergooding I would not trust other parts of this code either. For instance, `SetUInt32` and `SetUInt64` allow creating non-canonical zero values with `_length` greater than zero. That means `IsZero` may not work correctly.

tannergooding commented:
> For instance, SetUInt32 and SetUInt64 allow creating non-canonical zero values with _length greater than zero.

Yes, but it should never actually happen given the existing usages and if it did; would just result in a slower computation; not necessarily an incorrect computation. That is, the code as is being used; and as was ported from the cpp code; is still functioning correctly.

There are certainly cases where `Debug.Assert` could be added and where the code is needlessly complicated (as a given scenario will never be hit). So, I'm just saying it would likely be better to simplify the code as part of fixing those; rather than trying to fix the logic to cover a case that will never be hit.

AntonLapounov commented:
> not necessarily an incorrect computation

Well, `Compare` is also affected and we use it for control flow.  For example:
``` C#
var x = new BigInteger();
x.SetUInt64(0);
var y = new BigInteger(0);

// Outputs 1
Console.WriteLine(BigInteger.Compare(ref x, ref y));
```
I agree that if there were an easy way to hit one of these bugs, we would have hit it a long time ago.  Still that makes it very challenging to reason about this code.

AntonLapounov commented:
@ts2do Thank you for spotting these bugs.  As @tannergooding mentioned, the best approach is to remove both problematic methods by inlining them into their callers and simplifying.  Please let us know if you might help with that or prefer us to change the code.

ts2do commented:
The static Add method overload is indeed being used directly in src/System.Private.CoreLib/shared/System/Number.Dragon4.cs.
As for ShiftLeft, wouldn't it be better to fix the bug to simply support arbitrary shifting instead of limiting the shift parameter to 32?

ts2do commented:
I've made the requested changes.

I also have a more heavily modified version of BigInteger that I believe make it more consistent (e.g., results are always assigned to out variables), make it safer (e.g., added Debug.Assert calls which ensure that the result argument for Multiply and DivRem does not share the same address as one of the operands), and make it more correct (e.g., skip Memcpy if rhs is the same address as this). I figured it would be best to feed these critical bug fixes instead of giving it all at once and having them lost in the noise.

Here's the a gist with all of the changes:
https://gist.github.com/ts2do/629f3bc1a92eb10be18c1898b6169094

A few more notes:
I removed the `BigInteger(uint)` and `BigInteger(ulong)` constructors, as benchmarks results compared against directly setting fields have me thinking the fixed array is being zeroed. They would be replaced by `static void SetUInt32(out BigInteger result, uint value)` and `static void SetUInt32(out BigInteger result, ulong value)` to avoid that behavior.
I added an `AsString` method to BigInteger to help me debug it a little.

AntonLapounov commented:
@ts2do Note that Dragon4 uses _different_ `ShiftLeft` and `Add` overloads.  The two problematic overloads have only a single caller each, which are in the same Number.BigInteger.cs file. In particular, `ShiftLeft` is used for `Pow2` only.  We want to remove these problematic methods, not to fix them.  There are more issues, even with your latest commit.

ts2do commented:
My mistake, I was searching some local changes that I was toying around with. Sorry for the delay, I've been having some trouble running tests against the changes (though I got it building). I plan to redo my setup with a fresh installs soon, so until then, I guess I'll use CI to test it?

AntonLapounov commented:
@ts2do Thank you. Your changes look great in general. Reviewing now.

AntonLapounov commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -1065,29 +989,35 @@ private static uint SubtractDivisor(ref BigInteger lhs, int lhsStartIndex, ref B
+                        // No bit to carry
+                        return;
+                    }
+                }
```
You can simplify overflow checking:
``` C#
_blocks[0] += value;
if (_blocks[0] >= value)
{
    // No carry
    return;
}

for (int index = 1; index < length; index++)
{
    _blocks[index]++;
    if (_blocks[index] > 0)
    {
        // No carry
        return;
    }
}
```

AntonLapounov commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -923,58 +898,7 @@ public static void Pow10(uint exponent, out BigInteger result)
-            }
-
-            private static uint AddDivisor(ref BigInteger lhs, int lhsStartIndex, ref BigInteger rhs)
+            private static unsafe uint AddDivisor(ref BigInteger lhs, int lhsStartIndex, ref BigInteger rhs)
```
Merge conflict: please remove `unsafe`.

AntonLapounov commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -851,8 +820,14 @@ public static void Multiply(ref BigInteger lhs, ref BigInteger rhs, ref BigInteg
+                Debug.Assert(unchecked((uint)result._length) <= MaxBlockCount);
+                if (blocksToShift > 0)
+                {
+                    Buffer.ZeroMemory((byte*)(result.GetBlocksPointer()), ((blocksToShift) * sizeof(uint)));
```
We are trying to remove superfluous parentheses across CoreLib:
``` C#
Buffer.ZeroMemory((byte*)result.GetBlocksPointer(), blocksToShift * sizeof(uint));
```

AntonLapounov commented:
@tannergooding Don't we have a real product bug in `Multiply(ref, uint, ref)`?  We clearly call that method with `lhs` ≠`result` in four places in `Dragon4`.  For instance:
``` C#
var x = new BigInteger(7);
var y = new BigInteger(0);
BigInteger.Multiply(ref x, 6, ref y);
// Expected value: 42, actual value: 0
Console.WriteLine(y.ToUInt64());
```

AntonLapounov commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -771,6 +736,10 @@ public static void Multiply(ref BigInteger lhs, uint value, ref BigInteger resul
                 }
+                else
+                {
+                    result._length = lhsLength;
```
Nice catch!

tannergooding commented:
> Don't we have a real product bug in Multiply(ref, uint, ref)

There is a bug, but I don't believe it is one that can repro in production. The usages in `Number.Dragon4` are around `scaledMarginLow` and `pScaledMarginHigh` and we only call `Multiply` when `pScaledMarginHigh != &scaledMarginLow`. Due to how the code works, the length of `pScaledMarginHigh` will always be greater than `pScaledMarginHigh` and will either stay the same (no carry) or will grow by one (in which case we do update it).

AntonLapounov commented:
> the length of `pScaledMarginHigh` will always be greater than or equal to `scaledMarginLow`

@tannergooding What about this code path, where we multiply `scaledMarginLow` by a presumably big power of 10, which may make its length greater than the length of `pScaledMarginHigh`?
https://github.com/dotnet/coreclr/blob/5b1c001bc0fb5bb1ce035c02ec275763f66defc8/src/System.Private.CoreLib/shared/System/Number.Dragon4.cs#L222-L226

tannergooding commented:
> What about this code path, where we multiply scaledMarginLow by a presumably big power of 10, which may make its length greater than the length of pScaledMarginHigh?

I'd need to check some math to determine if that is fine or not, but I believe it still works out due to how the margins exist and where they exist.

AntonLapounov commented:
@tannergooding I have tried
```C#
Dragon4Double(1.0 / (1L << 31), -1, true, ref buffer);
```
under a debugger and noticed that `*pScaledMarginHigh` was indeed calculated incorrectly as I anticipated. Namely, after multiplying by a power of 10, `scaledMarginLow` equals to 10⁹ and `*pScaledMarginHigh` must be 2×10⁹; however, its actual value is 2,820,130,816 due to not setting its `_length` field.  Then `Dragon4` uses that incorrect value in a loop.

After fixing `IsZero` calculation, some tests are failing with:
```
Process terminated. Assertion failed.
   at System.Number.NumberToFloatingPointBitsSlow(NumberBuffer& number, FloatingPointInfo& info, UInt32 positiveExponent, UInt32 integerDigitsPresent, UInt32 fractionalDigitsPresent) in /_/src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs:line 489
   at System.Number.NumberToFloatingPointBits(NumberBuffer& number, FloatingPointInfo& info) in /_/src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs:line 414
   at System.Number.NumberToDouble(NumberBuffer& number) in /_/src/System.Private.CoreLib/shared/System/Number.Parsing.cs:line 1996
```
Would you be able to look at them?

AntonLapounov commented:
It seems that before this fix `Dragon4Double` used to work incorrectly for 39% of the powers of two. Fortunately, we use this algorithm as a fallback only.

For instance, 2⁵⁵ is converted to `36028797018963968` before the fix and `36028797018963970` after the fix (one trailing digit shorter). 2⁹⁵⁷ is converted to `1.21816425142499988e+288` before the fix and `1.218164251425e+288`  after the fix (five trailing digits shorter).

tannergooding commented:
Looks like its failing because the `fractionalNumerator` is asserted to be `non-zero`: https://source.dot.net/#System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs,487

This assertion "should" still hold true as we have at least 1 fractional digit present

tannergooding commented:
Put up a PR for the fix here: https://github.com/dotnet/coreclr/pull/27688

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

tannergooding commented:
@ts2do, if you rebase your changes ontop of (or merge with) the latest master, everything should pass now and we can get this merged 😄

ts2do commented:
Everything should be all set.

AntonLapounov commented:
@tsdo Nice work — thanks a lot!
